pydantic/decorators.py

```diff
@@ -83,46 +88,109 @@ def __call__(
     _core_schema.FieldWrapValidatorFunction,
 ]
 
+V1RootValidator = Union[
+    _V1RootValidatorClsMethod,
+    _decorators.V1RootValidatorFunction,
+]
+
+
+# Allow both a V1 (assumed pre=False) or V2 (assumed mode='after') validator
+# We lie to type checkers and say we return the same thing we get
+# but in reality we return a proxy object that _mostly_ behaves like the wrapped thing
+_V1ValidatorType = TypeVar('_V1ValidatorType', bound=Union[V1Validator, 'classmethod[Any]', 'staticmethod[Any]'])
+_V2BeforeAfterOrPlainValidatorType = TypeVar(
+    '_V2BeforeAfterOrPlainValidatorType',
+    bound=Union[V2Validator, 'classmethod[Any]', 'staticmethod[Any]'],
+)
+_V2WrapValidatorType = TypeVar(
+    '_V2WrapValidatorType', bound=Union[V2WrapValidator, 'classmethod[Any]', 'staticmethod[Any]']
+)
+_V1RootValidatorFunctionType = TypeVar(
+    '_V1RootValidatorFunctionType',
+    bound=Union[
+        _decorators.V1RootValidatorFunction,
+        _V1RootValidatorClsMethod,
+        'classmethod[Any]',
+        'staticmethod[Any]',
+    ],
+)
+
 
-@overload
 def validator(
+    __field: str,
     *fields: str,
-    check_fields: bool | None = ...,
-    sub_path: tuple[str | int, ...] | None = ...,
+    pre: bool = False,
+    check_fields: bool | None = None,
     allow_reuse: bool = False,
-) -> Callable[[V2Validator | V1Validator], classmethod[Any]]:
-    ...
+    each_item: bool = False,
+) -> Callable[[_V1ValidatorType], _V1ValidatorType]:
+    """
+    Decorate methods on the class indicating that they should be used to validate fields
+    :param fields: which field(s) the method should be called on
+    :param sub_path: TODO
+    :param check_fields: whether to check that the fields actually exist on the model
+    :param allow_reuse: whether to track and raise an error if another validator refers to the decorated function
+    """
+    fields = tuple((__field, *fields))
+    warn(
+        'Pydantic V1 style `@validator` validators are deprecated.'
+        ' You should migrate to Pydantic V2 style `@field_validator` validators,'
+        ' see the migration guide for more details',
+        DeprecationWarning,
+        stacklevel=2,
+    )
+
+    mode: Literal['before', 'after'] = 'before' if pre is True else 'after'
+
+    def dec(f: Any) -> _decorators.PydanticDecoratorMarker[Any]:
+        if _decorators.is_instance_method_from_sig(f):
+            raise TypeError('`@validator` cannot be applied to instance methods')
```

Comment:
this TypeError is not covered by any tests

Reply:
added tests